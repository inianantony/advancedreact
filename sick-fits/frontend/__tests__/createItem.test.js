import { mount } from 'enzyme';
import wait from 'waait';
import Router from 'next/router';
import CreateItem, { CREATE_ITEM_MUTATION } from '../components/CreateItem';
import { MockedProvider } from 'react-apollo/test-utils';
import toJSON from 'enzyme-to-json';
import { fakeItem } from '../lib/testUtils';

const dogImage = "https://example.com/dog.jpg"
global.fetch = jest.fn().mockResolvedValue({
    json: () => ({
        secure_url: dogImage,
        eager: [{ secure_url: dogImage }],
    })
});

describe("<CreateItem/>", () => {
    it("renders and match snapshot", () => {
        const wrapper = mount(
            <MockedProvider>
                <CreateItem />
            </MockedProvider>
        );
        expect(toJSON(wrapper.find("form"))).toMatchSnapshot();
    });

    it("uploads a file when changed", async () => {
        const wrapper = mount(
            <MockedProvider>
                <CreateItem />
            </MockedProvider>
        );

        const input = wrapper.find("input[type='file']");
        input.simulate("change", { target: { files: ['fakedog.jpg'] } });
        await wait();
        const component = wrapper.find("CreateItem").instance();
        expect(component.state.image).toEqual(dogImage);
        expect(component.state.largeImage).toEqual(dogImage);
        expect(global.fetch).toHaveBeenCalled();
        global.fetch.mockReset();
    });

    it("handles state update", async () => {
        const wrapper = mount(
            <MockedProvider>
                <CreateItem />
            </MockedProvider>
        );

        wrapper.find("#title").simulate("change", { target: { value: 'Testing', name: "title" } });
        wrapper.find("#price").simulate("change", { target: { value: '50000', name: "price", type: 'number' } });
        wrapper.find("#description").simulate("change", { target: { value: 'A cool one', name: "description" } });
        await wait();
        const component = wrapper.find("CreateItem").instance();
        expect(component.state).toMatchSnapshot();
        global.fetch.mockReset();
    });

    it("creates an item when form is submitted", async () => {
        const item = fakeItem();
        const mocks = [{
            request: {
                query: CREATE_ITEM_MUTATION, variables: {
                    title: item.title,
                    description: item.description,
                    image: '',
                    largeImage: '',
                    price: item.price
                }
            },
            result: {
                data: {
                    createItem: {
                        ...item,
                        id: "abc123",
                        typename: 'Item',
                    }
                }
            }
        }];
        const wrapper = mount(
            <MockedProvider mocks={mocks}>
                <CreateItem />
            </MockedProvider>
        );

        wrapper.find("#title").simulate("change", { target: { value: item.title, name: "title" } });
        wrapper.find("#price").simulate("change", { target: { value: item.price, name: "price", type: 'number' } });
        wrapper.find("#description").simulate("change", { target: { value: item.description, name: "description" } });
        Router.router = {
            push: jest.fn()
        };
        wrapper.find("form").simulate('submit');
        await wait(50);
        expect(Router.router.push).toHaveBeenCalled();
        expect(Router.router.push).toHaveBeenCalledWith({ "pathname": "/item", "query": { "id": "abc123" } });
        global.fetch.mockReset();
    });
});