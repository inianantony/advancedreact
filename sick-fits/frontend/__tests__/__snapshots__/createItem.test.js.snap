// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CreateItem/> handles state update 1`] = `
Object {
  "description": "A cool one",
  "image": "",
  "largeImage": "",
  "price": "50000",
  "title": "Testing",
}
`;

exports[`<CreateItem/> renders and match snapshot 1`] = `
<form
  className="Form-sc-1xszr8q-0 krRxky"
  onSubmit={[Function]}
>
  <DisplayError
    error={Object {}}
  />
  <fieldset
    aria-busy={false}
    disabled={false}
  >
    <label
      htmlFor="file"
    >
      Image
      <input
        id="file"
        name="file"
        onChange={[Function]}
        placeholder="Upload an image"
        required={true}
        type="file"
      />
    </label>
    <label
      htmlFor="title"
    >
      Title
      <input
        id="title"
        name="title"
        onChange={[Function]}
        placeholder="Title"
        required={true}
        type="text"
        value=""
      />
    </label>
    <label
      htmlFor="price"
    >
      Price
      <input
        id="price"
        name="price"
        onChange={[Function]}
        placeholder="Price"
        required={true}
        type="number"
        value={0}
      />
    </label>
    <label
      htmlFor="description"
    >
      Description
      <textarea
        id="description"
        name="description"
        onChange={[Function]}
        placeholder="Enter a Description"
        required={true}
        type="description"
        value=""
      />
    </label>
    <button
      type="Submit"
    >
      Submit
    </button>
  </fieldset>
</form>
`;
